stdlib/src/memory/arc.mojo

```diff
@@ -45,32 +45,231 @@ print(Arc(String("ok"))[])
+
+        Returns:
+            True iff the strong ref count has been successfully incremented
+        """
+
+        # Why the loop?
+        # We want to ensure that we don't pretend
+        # to be a strong (and imply that payload is init'd)
+        # when we haven't shown to ourselves that we are one.
+        # This loop does the equivalent of
+        # `fetch_add_if_neq(add: 1, if_neq: 0)`
+        @parameter
+        if enable_weak:
+            while True:
+                var cur_count = self.refcount.load()
+                if cur_count == 0:
+                    return False
+
+                var success = self.refcount.compare_exchange_weak(
+                    cur_count, cur_count + 1
+                )
+
+                # TODO: should this be marked `unlikely` once we have intrinsic for it?
+                # This only occurs during races on `refcount` when the strong refs
+                # have churn.
+                if not success:
+                    continue
+
+                return True
+        else:
+            return False
 
     fn add_ref(inout self):
         """Atomically increment the refcount."""
         _ = self.refcount.fetch_add(1)
 
+        # any strong is also, underneath, a weak
+        @parameter
+        if enable_weak:
+            self.add_weak()
+
+    fn add_weak(inout self):
+        """Atomically increment the weakref count"""
+        @parameter
+        if enable_weak:
+            _ = self._weakref_ptr()[].fetch_add(1)
+
     fn drop_ref(inout self) -> Bool:
-        """Atomically decrement the refcount and return true if the result
-        hits zero."""
-        return self.refcount.fetch_sub(1) == 1
+        """
+        Atomically decrement the refcount and weakref count.
+        Returns:
+            True iff the backing allocation for `self`
+            should (must!) be deleted
+        """
+        var last_strong = self.refcount.fetch_sub(1) == 1
+
+        if last_strong:
+            self._destroy_value()
+
+        # defer to drop_weak() for whether the backing
+        # mem should be dealloc'd
+        @parameter
+        if enable_weak:
+            return self.drop_weak()
+        else:
+            return last_strong
+
+    fn drop_weak(inout self) -> Bool:
+        """
+        Atomically decrement the weakref count, and return
+        true if the result hits zero
+        Returns:
+            True iff the backing allocation for `self`
+            should (must!) be deleted
+        """
+        @parameter
+        if enable_weak:
+            var last_any = self._weakref_ptr()[].fetch_sub(1) == 1
+
+            # if this is the last weak, and there
+            # are no other strongs, then we can dealloc the inner
+            return last_any
+        else:
+            #TODO: statically ensure unreachable?
+            return False
+
+
+@register_passable
+struct Weak[T: Movable, enable_weak: Bool = True](CollectionElement, CollectionElementNew):
```

Comment:
**Suggestion** I think in principle it should be possible for us to remove this `enable_weak` parameter from `Weak` itself.

The logic being that an instance of `Weak` can only have been derived from an `Arc[T, True]`, so the parameter would always have to be `True` anyway.

Reply:
I think this is an artifact of having written the first pass when cond-conf was...rather "newly minted" and had some rather squirrely behavior. I can see if I can pull it out now without knocking over the jenga tower.